Require fails when using ChromeDriver and HTTP url

Requires seems to fail (sometimes intermittently) when using ChromeDriver to launch an app that uses an HTTP URL. This only seems to be an issue with `http://` urls, `file://` urls do not exhibit this behavior.

To reproduce:

- Clone https://gist.github.com/kevinsawicki/a7f87097a8edcdfb43b7
- Run `npm install`
- Run `chromedriver --url-base=wd/hub --port=9515`
- Run `./run.sh` to see it launch normally
- Run `./test.sh` to see if fail to launch when run via ChromeDriver

The failure looks like:

<img width="641" alt="screen shot 2015-12-04 at 2 15 53 pm" src="https://cloud.githubusercontent.com/assets/671378/11602756/8ef2d860-9a91-11e5-919a-cfd43e161681.png">

The sample project uses `webdriverio` to launch the app.

The `index.html` is really simple and just does a `require` after the document is ready.

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <script type="text/javascript">
      function requireElectron() {
        try {
          require('electron')
          document.write('require(\'electron\') succeeded')
        } catch (error) {
          document.write('require(\'electron\') failed')
          document.write('<pre>' + error.stack + '</pre>')
        }
      }
    </script>
  </head>
  <body onload="requireElectron()">
  </body>
</html>
```

/cc @mkscrg

Hm, did some more debugging. I can confirm what zcbenz is saying.

```Destroy``` in atom_api_asar.cc is being called, which deconstructs the archive.
Later on when require is called ,```Stat`` in atom_api_asar.cc is called. Since the  ```archive_``` address is now null (0x00), Stat returns false.

I added some LOG statements to electron's source code to find this out.

The below is a big cryptic, sorry..
```
 []67788:1230/103843:ERROR:atom_api_asar.cc(55)] atom_api_asar STAT archive renderer/lib/web-view/web-view.js 0x7fa088cd5150 0x7fa088ccffe0
 []67788:1230/103843:ERROR:archive.cc(203)] Statting THIS PATH, FD 0x7fa088ccffe0 renderer/lib/web-view/web-view.js , 31
 []67788:1230/103843:ERROR:atom_api_asar.cc(55)] atom_api_asar STAT archive  0x7fa088cd5150 0x7fa088ccffe0
*[]67788:1230/103843:ERROR:archive.cc(203)] Statting THIS PATH, FD 0x7fa088ccffe0  , 31
*[]67788:1230/103843:ERROR:atom_api_asar.cc(105)] atom_api_asar DESTROYING ARCHIVE0x7fa088cd5150 0x7fa088ccffe0
*[]67788:1230/103843:ERROR:archive.cc(133)] Archive is deconstructing! THIS FD0x7fa088ccffe0 31
*[]67780:1230/103844:INFO:CONSOLE(21081)] "Got to require", source: file:///Users/justintaft/programming/LightTable/deploy/core/node_modules/lighttable/bootstrap.js (21081)
 []67780:1230/103844:INFO:CONSOLE(21082)] "1451500724137", source: file:///Users/justintaft/programming/LightTable/deploy/core/node_modules/lighttable/bootstrap.js (21082)
*[]67788:1230/103844:ERROR:atom_api_asar.cc(55)] atom_api_asar STAT archive common/api/lib 0x7fa088cd5150 0x0
*[]67788:1230/103844:ERROR:atom_api_asar.cc(55)] atom_api_asar STAT archive renderer/api/lib 0x7fa088cd5150 0x0
*[]67788:1230/103844:ERROR:atom_api_asar.cc(55)] atom_api_asar STAT archive renderer/api/lib/exports 0x7fa088cd5150 0x0
 []67780:1230/103844:INFO:CONSOLE(338)] "Uncaught Error: Cannot find module 'clipboard'", source: module.js (338)
 []67780:1230/103844:INFO:CONSOLE(36)] "TypeError: Cannot read property 'init' of undefined", source: file:///Users/justintaft/programming/LightTable/deploy/core/LightTable.html?id=1 (36)
 []67780:1230/103844:INFO:CONSOLE(37)] "TypeError: Cannot read property 'init' of undefined
     at HTMLScriptElement.script.onload (file:///Users/justintaft/programming/LightTable/deploy/core/LightTable.html?id=1:34:28)", source: file:///Users/justintaft/programming/LightTable/deploy/core/LightTable.html?id=1 (37)
```

--- a/atom/renderer/atom_renderer_client.cc
+++ b/atom/renderer/atom_renderer_client.cc
@@ -52,9 +52,14 @@ class AtomRenderFrameObserver : public content::RenderFrameObserver {
   // content::RenderFrameObserver:
   void DidCreateScriptContext(v8::Handle<v8::Context> context,
                               int extension_group,
-                              int world_id) {
+                              int world_id) override {
     renderer_client_->DidCreateScriptContext(
         render_frame()->GetWebFrame(), context);
+  }
+
+  void WillReleaseScriptContext(v8::Local<v8::Context> context,
+                                int world_id) override {
+    renderer_client_->WillReleaseScriptContext(context);
   }
 
  private:
@@ -160,6 +165,10 @@ void AtomRendererClient::DidCreateScriptContext(
 
   // Load everything.
   node_bindings_->LoadEnvironment(env);
+}
+
+void AtomRendererClient::WillReleaseScriptContext(
+    v8::Handle<v8::Context> context) {
 }
 
 bool AtomRendererClient::ShouldFork(blink::WebLocalFrame* frame,

--- a/atom/renderer/atom_renderer_client.cc
+++ b/atom/renderer/atom_renderer_client.cc
@@ -55,7 +55,7 @@ class AtomRenderFrameObserver : public content::RenderFrameObserver {
   void DidCreateScriptContext(v8::Handle<v8::Context> context,
                               int extension_group,
                               int world_id) override {
-    if (world_id_ != -1)
+    if (world_id_ != -1 && world_id_ != world_id)
       return;
     world_id_ = world_id;
     renderer_client_->DidCreateScriptContext(context);
@@ -88,9 +88,6 @@ void AtomRendererClient::WebKitInitialized() {
   blink::WebCustomElement::addEmbedderCustomElementName("webview");
   blink::WebCustomElement::addEmbedderCustomElementName("browserplugin");
 
-  // Allow file scheme to handle service worker by default.
-  blink::WebSecurityPolicy::registerURLSchemeAsAllowingServiceWorkers("file");
-
   OverrideNodeArrayBuffer();
 
   node_bindings_->Initialize();
@@ -122,6 +119,9 @@ void AtomRendererClient::RenderThreadStarted() {
 void AtomRendererClient::RenderFrameCreated(
     content::RenderFrame* render_frame) {
   new PepperHelper(render_frame);
+
+  // Allow file scheme to handle service worker by default.
+  blink::WebSecurityPolicy::registerURLSchemeAsAllowingServiceWorkers("file");
 
   // Only insert node integration for the main frame.
   if (!render_frame->IsMainFrame())

--- a/atom/renderer/atom_renderer_client.cc
+++ b/atom/renderer/atom_renderer_client.cc
@@ -47,22 +47,27 @@ class AtomRenderFrameObserver : public content::RenderFrameObserver {
   AtomRenderFrameObserver(content::RenderFrame* frame,
                           AtomRendererClient* renderer_client)
       : content::RenderFrameObserver(frame),
+        world_id_(-1),
         renderer_client_(renderer_client) {}
 
   // content::RenderFrameObserver:
   void DidCreateScriptContext(v8::Handle<v8::Context> context,
                               int extension_group,
                               int world_id) override {
-    renderer_client_->DidCreateScriptContext(
-        render_frame()->GetWebFrame(), context);
+    if (world_id_ != -1)
+      return;
+    world_id_ = world_id;
+    renderer_client_->DidCreateScriptContext(context);
   }
-
   void WillReleaseScriptContext(v8::Local<v8::Context> context,
                                 int world_id) override {
+    if (world_id_ != world_id)
+      return;
     renderer_client_->WillReleaseScriptContext(context);
   }
 
  private:
+  int world_id_;
   AtomRendererClient* renderer_client_;
 
   DISALLOW_COPY_AND_ASSIGN(AtomRenderFrameObserver);
@@ -81,6 +86,9 @@ AtomRendererClient::~AtomRendererClient() {
 void AtomRendererClient::WebKitInitialized() {
   blink::WebCustomElement::addEmbedderCustomElementName("webview");
   blink::WebCustomElement::addEmbedderCustomElementName("browserplugin");
+
+  // Allow file scheme to handle service worker by default.
+  blink::WebSecurityPolicy::registerURLSchemeAsAllowingServiceWorkers("file");
 
   OverrideNodeArrayBuffer();
 
@@ -113,10 +121,12 @@ void AtomRendererClient::RenderThreadStarted() {
 void AtomRendererClient::RenderFrameCreated(
     content::RenderFrame* render_frame) {
   new PepperHelper(render_frame);
-  new AtomRenderFrameObserver(render_frame, this);
 
-  // Allow file scheme to handle service worker by default.
-  blink::WebSecurityPolicy::registerURLSchemeAsAllowingServiceWorkers("file");
+  // Only insert node integration for the main frame.
+  if (!render_frame->IsMainFrame())
+    return;
+
+  new AtomRenderFrameObserver(render_frame, this);
 }
 
 void AtomRendererClient::RenderViewCreated(content::RenderView* render_view) {
@@ -144,12 +154,7 @@ bool AtomRendererClient::OverrideCreatePlugin(
 }
 
 void AtomRendererClient::DidCreateScriptContext(
-    blink::WebFrame* frame,
     v8::Handle<v8::Context> context) {
-  // Only insert node integration for the main frame.
-  if (frame->parent())
-    return;
-
   // Give the node loop a run to make sure everything is ready.
   node_bindings_->RunMessageLoop();
 

--- a/atom/renderer/atom_renderer_client.cc
+++ b/atom/renderer/atom_renderer_client.cc
@@ -9,6 +9,7 @@
 
 #include "atom/common/api/api_messages.h"
 #include "atom/common/api/atom_bindings.h"
+#include "atom/common/api/event_emitter_caller.h"
 #include "atom/common/node_bindings.h"
 #include "atom/common/node_includes.h"
 #include "atom/common/options_switches.h"
@@ -174,6 +175,8 @@ void AtomRendererClient::DidCreateScriptContext(
 
 void AtomRendererClient::WillReleaseScriptContext(
     v8::Handle<v8::Context> context) {
+  node::Environment* env = node::Environment::GetCurrent(context);
+  mate::EmitEvent(env->isolate(), env->process_object(), "exit");
 }
 
 bool AtomRendererClient::ShouldFork(blink::WebLocalFrame* frame,

--- a/atom/renderer/atom_renderer_client.h
+++ b/atom/renderer/atom_renderer_client.h
@@ -22,8 +22,7 @@ class AtomRendererClient : public content::ContentRendererClient,
   AtomRendererClient();
   virtual ~AtomRendererClient();
 
-  void DidCreateScriptContext(blink::WebFrame* frame,
-                              v8::Handle<v8::Context> context);
+  void DidCreateScriptContext(v8::Handle<v8::Context> context);
   void WillReleaseScriptContext(v8::Handle<v8::Context> context);
 
  private:

--- a/atom/renderer/atom_renderer_client.h
+++ b/atom/renderer/atom_renderer_client.h
@@ -24,6 +24,7 @@ class AtomRendererClient : public content::ContentRendererClient,
 
   void DidCreateScriptContext(blink::WebFrame* frame,
                               v8::Handle<v8::Context> context);
+  void WillReleaseScriptContext(v8::Handle<v8::Context> context);
 
  private:
   enum NodeIntegration {

--- a/atom/renderer/lib/init.js
+++ b/atom/renderer/lib/init.js
@@ -106,11 +106,6 @@ if (nodeIntegration === 'true' || nodeIntegration === 'all' || nodeIntegration =
       return false;
     }
   };
-
-  // Emit the 'exit' event when page is unloading.
-  window.addEventListener('unload', function() {
-    return process.emit('exit');
-  });
 } else {
   // Delete Node's symbols after the Environment has been loaded.
   process.once('loaded', function() {

